taskflows/openchemistry/__init__.py

```diff
@@ -0,0 +1,449 @@
+from bson.objectid import ObjectId
+import jsonpickle
+import cumulus
+from cumulus.celery import command, monitor
+from cumulus.taskflow import TaskFlow
+from cumulus.taskflow.cluster import create_girder_client
+from cumulus.tasks.job import (download_job_input_folders,
+                               upload_job_output_to_folder)
+from cumulus.tasks.job import submit_job, monitor_job
+
+from girder.api.rest import getCurrentUser
+from girder.constants import AccessType
+from girder.utility.model_importer import ModelImporter
+
+from jsonpath_rw import parse
+import os
+import datetime
+import json
+from io import BytesIO
+import tempfile
+
+from abc import ABC, abstractmethod
+
+command.conf.update(
+    CELERY_TASK_SERIALIZER='jsonpickle'
+)
+
+monitor.conf.update(
+    CELERY_TASK_SERIALIZER='jsonpickle'
+)
+
+class OpenChemistryTaskFlow(TaskFlow, ABC):
+    """
+    {
+        "input": {
+            "calculation": {
+                "_id": <the id of the pending calculation>
+            },
+        },
+        "cluster": {
+            "_id": <id of cluster to run on>
+        }
+    }
+    """
+
+    input_name = 'oc.input'
+    job_name = 'oc_run'
+    logger_name = 'Create OpenChemistry job.'
+
+    def start(self, *args, **kwargs):
+        user = getCurrentUser()
+        input_ = kwargs.get('input')
+        cluster = kwargs.get('cluster')
+
+        if input_ is None:
+            raise Exception('Unable to extract input.')
+
+        if '_id' not in cluster and 'name' not in cluster:
+            raise Exception('Unable to extract cluster.')
+
+        cluster_id = parse('cluster._id').find(kwargs)
+        if cluster_id:
+            cluster_id = cluster_id[0].value
+            model = ModelImporter.model('cluster', 'cumulus')
+            cluster = model.load(cluster_id, user=user, level=AccessType.ADMIN)
+            cluster = model.filter(cluster, user, passphrase=False)
+
+        super(OpenChemistryTaskFlow, self).start(
+            setup_input_template.s(input_, cluster, self.__class__),
+            *args, **kwargs)
+
+    @staticmethod
+    @abstractmethod
+    def input_generator(params, tmp_file):
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def copy_output_files(filenames):
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def ec2_job_commands(input_name):
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def demo_job_commands(input_name):
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def nersc_job_commands(input_name):
+        pass
```

Comment:
May be not in the PR but we should abstract away the differences between these environments. I think Singularity can help use here, as we would then be able compile our docker containers into images that we would use at NERSC.